File: 383_ransom_note/canConstructArray.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        BOUND = ord('a')

        if len(ransomNote) > len(magazine):
            return False
        
        counter = [0] * 26

        for i in magazine:
            counter[ord(i) - BOUND] += 1

        for i in ransomNote:
            if counter[ord(i) - BOUND] == 0:
                return False
            counter[ord(i) - BOUND] -= 1
            
        return True

File: 383_ransom_note/canConstructArray.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        int n = magazine.length();
        int m = ransomNote.length();
        if (m > n) {
            return false;
        }

        int[] counter = new int[26];

        for (int i = 0; i < n; i++) {
            char c = magazine.charAt(i);
            counter[c - 'a']++;
        }

        for (int i = 0; i < m; i++) {
            char c = ransomNote.charAt(i);
            if (counter[c - 'a'] == 0) {
                return false;
            }
            counter[c - 'a']--;
        }

        return true;
    }
}